### PR TITLE
[WIP] Accelerated rendering

### PR DESCRIPTION
**Pull request by Vultraz**
WIP work to get the game using textures and/or accelerated rendering. Opening a PR for feedback as I go along. GUI2 almost entirely works except for some alpha issues with the text in the story viewer and the minimap not showing up.

### CODE REVIEW

**Comment by Vultraz** [comment]
@aginor I know you've ceased development for the project, I wouldn't mind some general feedback from you as you're good with SDL (if you feel like it, that is).

**Comment by CelticMinstrel** [comment]
Just make sure you _don't merge this_.

**Comment by Vultraz** [comment]
I am attempting to get this to a mergable state. It is not there yet.

**Comment by CelticMinstrel** [comment]
There is no mergeable state that will be acceptable for this PR in the near future.

**Comment by Vultraz** [comment]
Well, not exactly true, not if I can solve the black screen issue when gui2 dialogs appear in-game.

**Comment by CelticMinstrel** [comment]
You are still completely missing the point. This is a big change and/or refactor. We are in the beginning of a release cycle. Do you intend to release 1.14 this year or not?

**Comment by Vultraz** [comment]
Of course.

File: src/sdl/window.cpp [hunk 1]
```diff
@@ -54,6 +54,12 @@ window::window(const std::string& title,
 	// Set default blend mode to blend.
 	SDL_SetRenderDrawBlendMode(*this, SDL_BLENDMODE_BLEND);
 
+	// Use linear scaling when rendering, if applicable.
+	SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "linear");
+
+	// Enable 3D rendering for the window surface. Not sure if this really does anything.
+	SDL_SetHint(SDL_HINT_FRAMEBUFFER_ACCELERATION, "1");
```
Comment (aginor) [inline src/sdl/window.cpp]: This is setting the defaults, you're better off omitting it as you may break things by trying to override it on an unsupported platform.

File: src/gui/core/canvas_private.hpp [hunk 2]
```diff
@@ -252,9 +260,11 @@ class text_shape : public canvas::shape {
 	explicit text_shape(const config& cfg);
 
 	/** Implement shape::draw(). */
-	void draw(surface& canvas,
-			  SDL_Renderer* renderer,
-			  wfl::map_formula_callable& variables) override;
+	void draw(
+			const int canvas_w,
+			const int canvas_h,
+			SDL_Renderer* renderer,
```
Comment (aginor) [inline src/gui/core/canvas_private.hpp]: Don't pass raw pointers

File: src/gui/core/canvas_private.hpp [hunk 3]
```diff
@@ -186,10 +193,11 @@ class image_shape : public canvas::shape {
...
 
 	/** Implement shape::draw(). */
-	void draw(surface& canvas,
-			  SDL_Renderer* renderer,
-			  wfl::map_formula_callable& variables) override;
-
+	void draw(
+			const int canvas_w,
+			const int canvas_h,
+			SDL_Renderer* renderer,
```
Comment (aginor) [inline src/gui/core/canvas_private.hpp]: Don't pass raw pointers

File: src/gui/core/canvas_private.hpp [hunk 4]
```diff
@@ -157,9 +162,11 @@ class circle_shape : public canvas::shape {
 	explicit circle_shape(const config& cfg);
 
 	/** Implement shape::draw(). */
-	void draw(surface& canvas,
-			  SDL_Renderer* renderer,
-			  wfl::map_formula_callable& variables) override;
+	void draw(
+			const int canvas_w,
+			const int canvas_h,
+			SDL_Renderer* renderer,
```
Comment (aginor) [inline src/gui/core/canvas_private.hpp]: Don't pass raw pointers

File: src/gui/core/canvas.hpp [hunk 5]
```diff
@@ -196,9 +189,10 @@ class canvas
 	/** Height of the canvas. */
 	unsigned h_;
 
-	/** The surface we draw all items on. */
-	surface canvas_;
+	/** The texture onto which items are drawn. */
+	SDL_Texture* texture_;
```
Comment (aginor) [inline src/gui/core/canvas.hpp]: This should be a managed class, not a raw pointer.

File: src/gui/core/canvas.cpp [hunk 6]
```diff
@@ -1121,6 +1120,7 @@ void image_shape::draw(surface& canvas,
 				DBG_GUI_D << "Image: vertical stretch from " << image_->w << ','
 						  << image_->h << " to a height of " << h << ".\n";
 
+				// TODO: convert to texture handling.
```
Comment (aginor) [inline src/gui/core/canvas.cpp]: you just render it and let the renderer do the stretching for you

**Comment by aginor** [comment]
It might also be worth mentioning that I cannot test this branch as it segfaults immediately.

```
#0  0x0000000001180ea7 in screen_area() ()
#1  0x0000000000dda3cb in gui2::window::update_screen_size() [clone .part.38] ()
#2  0x0000000000d940de in gui2::load_settings() ()
#3  0x0000000000d4e799 in gui2::init() ()
#4  0x0000000000843f08 in main ()
```

**Comment by Vultraz** [comment]
Oh dear. That's not good. O_O

File: src/sdl/window.cpp (hunk 1 above)
Comment (Vultraz) [inline src/sdl/window.cpp]: Don't hints get ignored if unsupported?

**Comment by aginor** [comment]
I never enabled the framebuffer acceleration flags because I found in my testing that they caused a lot of problems across different platforms (nothing insurmountable), but fixing it all would take a lot of time and testing would be difficult. You will see different behaviours across different operating systems and between different hardware configurations. Most of the difficulty will be in setting it all up properly, once that's done SDL should go and abstract it all away for you.

**Comment by Vultraz** [comment]
Hm. I'll disable, it, then.

File: src/gui/core/canvas.hpp (hunk 5 above)
Comment (Vultraz) [inline src/gui/core/canvas.hpp]: Working on it as we speak.

**Comment by Vultraz** [comment]
@aginor I've added the wrapper class.

**Comment by GregoryLundberg** [comment]
You can't just change the copyright from 2003-2017 to simply read 2017.

You're fine with just 2017 on a completely new file or a completely rewritten file. But if you're changing a file all you can do is extend the range to include 2017. The parts you didn't change remain copyright 2003 (or whatever).

**Comment by Vultraz** [comment]
@GregoryLundberg it's a new file I wrote that I had just added eaelier where I forgot to remove the '2003' part from the copyright when I copied it (the copyright) from another file.

**Comment by Vultraz** [comment]
@aginor the startup crash should be fixed.

**Comment by Vultraz** [comment]
So, just want to leave a progress report. As you can see in the above commits, I've made per-frame drawing global and am now working to refactor the main display to properly work with this drawing method. This is what I have so far:
progress2

**Comment by Vultraz** [comment]
@aginor especially could use feedback on b7000a5. Not sure if I'm supposed to be calling `join()` outside the ctor.

**Comment by Vultraz** [comment]
@aginor actually f50c077 now.

**Comment by CelticMinstrel** [comment]
I think it would've actually been better if that Xcode project fix _wasn't_ merged, since this branch now has a conflict in the Xcode project that will need to be resolved when it's rebased...

**Comment by Vultraz** [comment]
I thought I removed that commit....

**Comment by CelticMinstrel** [comment]
Which one?

**Comment by Vultraz** [comment]
Merged manually.

**Pull request closed without merging**